client/consensus/manual-seal/src/seal_new_block.rs

```diff
@@ -0,0 +1,165 @@
+	pub select_chain: &'a C,
+	/// block import object
+	pub block_import: &'a mut BoxBlockImport<B>,
+	/// inherent data provider
+	pub inherent_data_provider: &'a InherentDataProviders,
+}
+
+/// seals a new block with the given params
+pub async fn seal_new_block<B, C, CB, E, P, H>(params: SealBlockParams<'_, B, C, CB, E, P>)
+	where
+		B: BlockT,
+		H: Hasher<Out=<B as BlockT>::Hash>,
+		CB: ClientBackend<B, H>,
+		E: Environment<B>,
+		E::Error: std::fmt::Display,
+		<E::Proposer as Proposer<B>>::Error: std::fmt::Display,
+		P: txpool::ChainApi<Block=B, Hash=<B as BlockT>::Hash>,
+		C: SelectChain<B>,
+{
+	let SealBlockParams {
+		create_empty,
+		finalize,
+		pool,
+		parent_hash,
+		mut sender,
+		back_end,
+		select_chain,
+		block_import,
+		env,
+		inherent_data_provider,
+		..
+	} = params;
+
+	let future = async {
+		if pool.status().ready == 0 && !create_empty {
+			return Err(rpc::send_result(&mut sender, Err(Error::EmptyTransactionPool)));
+		}
+
+		// get the header to build this new block on.
+		// use the parent_hash supplied via `EngineCommand`
+		// or fetch the best_block.
+		let header = match parent_hash {
+			Some(hash) => {
+				back_end.blockchain().header(BlockId::Hash(hash))
+					.map_err(Error::from)
+					.and_then(|header| {
+						match header {
+							Some(header) => Ok(header),
+							None => Err(Error::BlockNotFound(format!("{}", hash)))
+						}
+					})
+			}
+			None => select_chain.best_chain().map_err(Error::from)
+		};
+
+		let header = header.map_err(|e| rpc::send_result(&mut sender, Err(e)))?;
+
+		let mut proposer = env.init(&header)
+			.map_err(|err| {
+				rpc::send_result(&mut sender, Err(Error::ProposerError(format!("{}", err))))
+			})?;
+
+		let id = inherent_data_provider.create_inherent_data()
+			.map_err(|err| {
+				rpc::send_result(&mut sender, Err(err.into()))
+			})?;
+
+		let block = proposer.propose(id, Default::default(), Duration::from_secs(5))
+			.await
+			.map_err(|err| {
+				rpc::send_result(&mut sender, Err(Error::ProposerError(format!("{}", err))))
+			})?;
+
+		if block.extrinsics().len() == 0 && !create_empty {
+			return Err(rpc::send_result(&mut sender, Err(Error::EmptyTransactionPool)))
+		}
+
+		let (header, body) = block.deconstruct();
+		let params = BlockImportParams {
+			origin: BlockOrigin::Own,
+			header: header.clone(),
+			justification: None,
+			post_digests: Vec::new(),
+			body: Some(body),
+			finalized: finalize,
+			auxiliary: Vec::new(),
+			fork_choice: ForkChoiceStrategy::LongestChain,
+			allow_missing_state: false,
+			import_existing: false,
+		};
+
+		match block_import.import_block(params, HashMap::new()) {
+			Ok(ImportResult::Imported(aux)) => {
+				rpc::send_result(&mut sender, Ok(CreatedBlock { hash: <B as BlockT>::Header::hash(&header), aux }))
+			}
+			Ok(other) => rpc::send_result(&mut sender, Err(other.into())),
+			Err(e) => {
+				log::warn!("Failed to import block: {:?}", e);
+				rpc::send_result(&mut sender, Err(e.into()))
```

Comment:
It's cool that now we have one line to send the result, but isn't it really possible to extract the logic and just have single place were we actually call that method? Doing `map_err` all over this function seems quite redundant.